device/transport/mqtt/src/mqtt_twin_client.ts

```diff
@@ -0,0 +1,266 @@
+ * @fires MqttTwinReceiver#error    an error has occured while subscribing to an MQTT topic
+ * @fires MqttTwinReceiver#response   a response message has been received from the service
+ * @fires MqttTwinReceiver#post a post message has been received from the service
+ * @throws {ReferenceError} If client parameter is falsy.
+ *
+ */
+export class MqttTwinClient extends EventEmitter {
+  static desiredPropertiesUpdateEvent: string = 'twinDesiredPropertiesUpdate';
+
+  private _mqtt: MqttBase;
+  private _pendingTwinRequests: { [key: string]: any } = {};
+  private _topicFsm: machina.BehavioralFsm;
+  private _responseTopic: TopicSubscription;
+  private _desiredPropertiesUpdatesTopic: TopicSubscription;
+
+  constructor(client: MqttBase) {
+    super();
+    /*Codes_SRS_NODE_DEVICE_MQTT_TWIN_CLIENT_18_001: [The `MqttTwinClient` constructor shall accept a `client` object.]*/
+    this._mqtt = client;
+
+    const messageHandler = this._onMqttMessage.bind(this);
+
+    /*Codes_SRS_NODE_DEVICE_MQTT_TWIN_CLIENT_16_001: [The `MqttTwinClient` constructor shall immediately subscribe to the `message` event of the `client` object.]*/
+    this._mqtt.on('message', messageHandler);
+
+    this._topicFsm = new machina.BehavioralFsm({
+      initialState: 'unsubscribed',
+      states: {
+        unsubscribed: {
+          _onEnter: function (client: TopicSubscription, err: Error, callback: (err?: Error) => void): void {
+            if (callback) {
+              callback(err);
+            }
+          },
+          subscribe: function (client: TopicSubscription, callback: (err?: Error) => void): void {
+            this.transition(client, 'subscribing', callback);
+          },
+          unsubscribe: function (client: TopicSubscription, callback: (err?: Error) => void): void {
+            // not entirely sure about that. if subscription are restored because cleanSession is false, it means technically a user may want to unsubscribe
+            // even though subscribe hasn't been called yet.
+            callback();
+          }
+        },
+        subscribing: {
+          _onEnter: function (client: TopicSubscription, callback: (err?: Error) => void): void {
+            client.mqttClient.subscribe(client.topic, { qos: 0 }, (err, result) => {
+              if (err) {
+                this.transition(client, 'unsubscribed', err, callback);
+              } else {
+                debug('subscribed to response topic: ' + JSON.stringify(result));
+                this.transition(client, 'subscribed', callback);
+              }
+            });
+          },
+          '*': function (client: TopicSubscription): void { this.deferUntilTransition(client); }
+        },
+        subscribed: {
+          _onEnter: function (client: TopicSubscription, callback: (err?: Error) => void): void {
+            callback();
+          },
+          subscribe: function (client: TopicSubscription, callback: (err?: Error) => void): void {
+            callback();
```

Comment:
>callback [](start = 12, length = 8)

Now I'm utterly confused.  If I'm reading this correctly, this object can only be subscribed to a single topic at a time, but it looks like there's only one and it handles both RES and PATCH topics.  What is this thing?  (still stuck on the fact that this is a client that takes one kind of client as a constructor arg, passes around a different kind of client that has a pointer to another client,  which comes back to this client.  < core dump>) :))

Reply:
see [http://machina-js.org/](http://machina-js.org/) in the Going Further - Behavioral FSM section. The client/topicSubscription object is used to store the state of the state machine. we have 1 single state machine definition used by 2 topics, independently. that's why we carry this object around.

---
In reply to: [170703083](https://github.com/Azure/azure-iot-sdk-node/pull/236#discussion_r170703083) [](ancestors = 170703083)